lib/commons/dom/create-grid.js

```diff
@@ -323,36 +317,100 @@ function findScrollRegionParent(vNode, parentVNode) {
  * @param {VirtualNode}
  */
 function addNodeToGrid(grid, vNode) {
-  const gridSize = constants.gridSize;
   vNode.clientRects.forEach(rect => {
-    if (rect.right <= 0 || rect.bottom <= 0) {
-      return;
-    }
     // save a reference to where this element is in the grid so we
     // can find it even if it's in a subgrid
     vNode._grid ??= grid;
-    const x = rect.left;
-    const y = rect.top;
+    const gridRect = grid.getGridPositionOfRect(rect);
+    grid.loopGridPosition(gridRect, gridCell => {
+      if (!gridCell.includes(vNode)) {
+        gridCell.push(vNode);
+      }
+    });
+  });
+}
 
-    // "| 0" is a faster way to do Math.floor
-    // @see https://jsperf.com/math-floor-vs-math-round-vs-parseint/152
-    const startRow = (y / gridSize) | 0;
-    const startCol = (x / gridSize) | 0;
-    const endRow = ((y + rect.height) / gridSize) | 0;
-    const endCol = ((x + rect.width) / gridSize) | 0;
+class Grid {
+  constructor(container = null) {
+    this.container = container;
+    this.cells = [];
+  }
 
-    grid.numCols = Math.max(grid.numCols ?? 0, endCol);
+  /**
+   * Convert x or y coordinate from rect, to a position in the grid
+   * @param {number}
+   * @returns {number}
+   */
+  toGridIndex(num) {
+    return Math.floor(num / constants.gridSize);
+  }
 
-    for (let row = startRow; row <= endRow; row++) {
-      grid.cells[row] = grid.cells[row] || [];
+  /**
+   * Return an an array of nodes available at a particular grid coordinate
+   * @param {DOMPoint} gridPosition
+   * @returns {Array<AbstractVirtualNode>|null}
+   */
+  getCellFromPoint({ x, y }) {
+    assert(this.boundaries, 'Grid does not have cells added');
+    const rowIndex = this.toGridIndex(y);
+    const colIndex = this.toGridIndex(x);
+    assert(
+      isPointInRect({ y: rowIndex, x: colIndex }, this.boundaries),
+      'Element midpoint exceeds the grid bounds'
+    );
+    const row = this.cells[rowIndex - this.cells._negativeIndex] ?? [];
+    return row[colIndex - row._negativeIndex] ?? null;
```

Comment:
Small thing. Maybe this should return an empty array instead of null? That way when you use it in get-rect-stack we wouldn't need to do default it to an empty array using `|| []` after the call?